Rewrite of test-suite

remen opened the pull request:
There was a submodule called "test-suite" containing a test of a pipeline. Great! However:
- It wasn't part of the parent pom build
- It downloaded and ran its own version of mongodb, which doesn't work on travis
- It didn't upload jar-with-dependencies to the core. Instead it ran stages "inline" (afaict)

I needed a full scale functional test suite while doing the refactoring, so I decided to rewrite it.

Commit:
integration_tests

Commit:
Merge branch 'travis_fixup' into integration_tests

remen commented:
Sorry about the `integration_tests` commit. Should have rebased before creating the pull request. Please let me know if you want me to redo it or if it's ok.

laserval commented:
Good stuff, although I believe it does something different than the old `test-suite`. That was built to test pipelines (you modify it with your pipeline). I haven't used it, but maybe @ssimon has. I think this is more useful to us though.

laserval commented on test-suite/functional-tests/src/test/java/com/findwise/hydra/FullScaleIT.java:
```diff
@@ -0,0 +1,197 @@
+	@Before
+	public void setUp() throws Exception {
+		mongoConfiguration = new MongoConfiguration();
+		mongoConnector = new MongoConnector(mongoConfiguration);
```
This test should use a non-default pipeline, like the integration tests in Core and the Mongo implementation.

Commit:
Adds functional test for Stage.init() and document failure

Commit:
Fixes Core configuration breaking when created with defaults progammatically

Commit:
Changes functional tests to use a non-default namespace. Removes unused configuration files.

laserval commented:
I've worked a bit with this now, and it's an excellent tool! We should add functional tests for some different scenarios to ensure we catch runtime-functionality. I suppose there are ways to improve it, but I think this is a great start.

I've created a pull request to your branch with an additional test (to catch that init() is run), as well as a bugfix and a change to use a non-default pipeline/namespace for the test.

remen commented:
I'm glad you like it. I think these types of tests are the ones that add the most value (of course you should _also_ write unit tests) and are most resilient to refactoring. I'll have a look at your changes, and might add some of my own  since I think I'm using stage groups wrong.

remen committed:
Merge pull request #2 from laserval/remen_integration_tests

Adds functional test for Stage.init() and changes to non-default namespace for test

remen commented:
Currently (because I wasn't thinking about it) there is one stage-group per stage. I don't really know what the "canonical" way of using Hydra is. What does the command line inserter do per default? What should we be testing?

laserval commented:
> What does the command line inserter do per default? What should we be testing?

The default is one group per stage. The only current benefit to stage groups is that you cut down on JVM overhead.
I think we should at least have both cases - i.e. we should have both a stage group with several stages and one with a single stage.

remen commented:
Good idea, I'll get right on it.

Commit:
Upgrade junit to 4.11 in functional-tests

Need it for @Parameters(name = ...)

Commit:
Added test in functional-tests for using a single stage group

- Now uses unique UUIDs as IDs to avoid tests interfering
- Extracted classes LinearPipelineBuilder and StageBuilder

remen commented:
Done. The change was quite invasive, but I think it makes it easier to understand the overall logic anyway.

Commit:
Fixup. Use assertThat instead of assertEquals
